File: hilos2.py
#El programa simula tareas con variacion de tiempos
import threading
import time
import random

# Variable global para rastrear el tiempo total compartido
tiempo_inicio = time.time()  # Guarda el momento en que comienzan los hilos

# Función que será ejecutada por los hilos
def imprimir_mensaje(nombre, veces):
    tiempo_individual = 0  # Contador para el tiempo individual del hilo
    for i in range(veces):
        tiempo_espera = random.uniform(0.5, 2.0)  # Genera un tiempo aleatorio entre 0.5 y 2.0 segundos
        time.sleep(tiempo_espera)  # Pausa por el tiempo aleatorio
        tiempo_individual += tiempo_espera  # Acumula el tiempo del hilo
        tiempo_total_compartido = time.time() - tiempo_inicio  # Calcula el tiempo total desde el inicio
        print(f"{nombre}: Iteración {i + 1}, Tiempo de espera: {tiempo_espera:.2f} s, "
              f"Tiempo individual: {tiempo_individual:.2f} s, Tiempo total compartido: {tiempo_total_compartido:.2f} s")

# Crear los hilos
hilo1 = threading.Thread(target=imprimir_mensaje, args=("Hilo 1", 3))
hilo2 = threading.Thread(target=imprimir_mensaje, args=("Hilo 2", 2))

# Iniciar los hilos
hilo1.start()
hilo2.start()

# Esperar a que los hilos terminen
hilo1.join()
hilo2.join()

# Imprimir mensaje final con el tiempo total compartido
tiempo_total = time.time() - tiempo_inicio
print(f"Todos los hilos han terminado. Tiempo total transcurrido: {tiempo_total:.2f} s.")

File: hilos2.java
package Java;
// Este programa simula tareas en hilos con variación en tiempos de espera
import java.util.Random;

class TareaConTiempo implements Runnable {
    private String nombre;
    private int veces;
    private long tiempoInicio;

    public TareaConTiempo(String nombre, int veces, long tiempoInicio) {
        this.nombre = nombre;
        this.veces = veces;
        this.tiempoInicio = tiempoInicio;
    }

    @Override
    public void run() {
        Random random = new Random();
        double tiempoIndividual = 0; // Contador para el tiempo individual del hilo

        for (int i = 0; i < veces; i++) {
            double tiempoEspera = 0.5 + (1.5 * random.nextDouble()); // Genera un tiempo aleatorio entre 0.5 y 2.0 segundos

            try {
                Thread.sleep((long) (tiempoEspera * 1000)); // Pausa por el tiempo aleatorio
            } catch (InterruptedException e) {
                System.out.println(nombre + " fue interrumpido.");
            }

            tiempoIndividual += tiempoEspera; // Acumula el tiempo del hilo
            double tiempoTotalCompartido = (System.currentTimeMillis() - tiempoInicio) / 1000.0; // Calcula el tiempo total desde el inicio

            System.out.printf("%s: Iteración %d, Tiempo de espera: %.2f s, Tiempo individual: %.2f s, Tiempo total compartido: %.2f s%n",
                    nombre, i + 1, tiempoEspera, tiempoIndividual, tiempoTotalCompartido);
        }
    }
}

public class hilos2{
    public static void main(String[] args) {
        long tiempoInicio = System.currentTimeMillis(); // Guarda el momento en que comienzan los hilos

        // Crear los hilos
        Thread hilo1 = new Thread(new TareaConTiempo("Hilo 1", 3, tiempoInicio));
        Thread hilo2 = new Thread(new TareaConTiempo("Hilo 2", 2, tiempoInicio));

        // Iniciar los hilos
        hilo1.start();
        hilo2.start();

        // Esperar a que los hilos terminen
        try {
            hilo1.join();
            hilo2.join();
        } catch (InterruptedException e) {
            System.out.println("La ejecución principal fue interrumpida.");
        }

        // Imprimir mensaje final con el tiempo total compartido
        double tiempoTotal = (System.currentTimeMillis() - tiempoInicio) / 1000.0;
        System.out.printf("Todos los hilos han terminado. Tiempo total transcurrido: %.2f s.%n", tiempoTotal);
    }
}
